src/main/java/io/reactivex/observers/SafeObserver.java

```diff
@@ -80,22 +77,60 @@ public void onNext(T t) {
         if (done) {
             return;
         }
-        if (t == null) {
-            onError(new NullPointerException());
+        if (s == null) {
+            onNextNoSubscription();
             return;
         }
-        if (s == null) {
-            onError(null); // null is okay here, onError checks for subscription == null first
+
+        if (t == null) {
+            Throwable ex = new NullPointerException();
+            try {
+                s.dispose();
+            } catch (Throwable e1) {
+                Exceptions.throwIfFatal(e1);
+                onError(new CompositeException(ex, e1));
+                return;
+            }
+            onError(ex);
             return;
         }
+        
         try {
             actual.onNext(t);
         } catch (Throwable e) {
             Exceptions.throwIfFatal(e);
+            try {
+                s.dispose();
+            } catch (Throwable e1) {
+                Exceptions.throwIfFatal(e1);
+                onError(new CompositeException(e, e1));
+                return;
+            }
             onError(e);
         }
     }
     
+    void onNextNoSubscription() {
+        
+        Throwable ex = new NullPointerException("Subscription not set!");
+        
+        try {
+            actual.onSubscribe(EmptyDisposable.INSTANCE);
+        } catch (Throwable e) {
+            Exceptions.throwIfFatal(e);
+            // can't call onError because the actual's state may be corrupt at this point
+            RxJavaPlugins.onError(new CompositeException(ex, e));
```

Comment:
First is the NPE, the second is the crash of the call; don't want to lose either or drift apart. Also naming is swapped sometimes but should be still in the correct order.